src/core/url-beautifier/search/generator.ts

```diff
@@ -1,60 +1,104 @@
-import { Store } from '@storefront/flux-capacitor';
-import { Request, SelectedRangeRefinement, SelectedRefinement, SelectedValueRefinement } from 'groupby-api';
+import { Adapters, Store } from '@storefront/flux-capacitor';
 import UrlBeautifier from '..';
 import { UrlGenerator } from '../handler';
+import * as utils from '../utils';
 
-export default class SearchUrlGenerator extends UrlGenerator<UrlBeautifier.SearchRequest> {
+export default class SearchUrlGenerator extends UrlGenerator<UrlBeautifier.UrlState> {
 
-  build = (request: UrlBeautifier.SearchRequest) => {
+  build = (state: UrlBeautifier.UrlState) => {
     const path = [];
     const query = {};
+    const initialState = Adapters.Configuration.initialState(this.beautifier.appConfig).data;
 
+    // TODO: fix destructive altering of state and remove layman's clone
     // layman's clone
-    request = { ...request, refinements: [...request.refinements] };
+    state = { ...state, refinements: [...state.refinements] };
 
-    if (request.query) {
-      path.push(request.query);
+    if (state.query) {
+      path.push(utils.encodeChars(state.query));
     }
 
     if (this.config.useReferenceKeys) {
-      path.push(...this.convertReferencedRefinements(request));
+      // TODO: prevent destructively altering state.refinements
+      path.push(...this.convertReferencedRefinements(state));
     } else {
-      path.push(...SearchUrlGenerator.convertPathRefinements(request));
+      path.push(...SearchUrlGenerator.convertPathRefinements(state.refinements));
+      // path.push('/');
     }
 
-    if (request.refinements.length !== 0) {
-      query[this.config.params.refinements] = SearchUrlGenerator.convertTrailingRefinements(request);
+    if (this.config.params.refinements && state.refinements.length !== 0) {
+      query[this.config.params.refinements] = utils.encodeArray(SearchUrlGenerator.sortRefinements(state));
     }
 
-    if ('pageSize' in request) {
-      query[this.config.params.pageSize] = request.pageSize;
+    if ('pageSize' in state) {
+      const initialSizes = initialState.page.sizes;
+      const initialPageSize = initialSizes.items[initialSizes.selected];
+      const pageSize = state.pageSize;
+      if (initialPageSize !== pageSize) {
+        query[this.config.params.pageSize] = pageSize;
+      }
+    }
+    if ('page' in state) {
+      const page = state.page;
+      if (initialState.page.first !== state.page)
```

Comment:
also, should we be checking if it equals `first`, or if it equals `current`? Just because `first` here refers to a record index, not a page. And I believe that the `current` should also be `1` initially